### PR TITLE
Fix extract block for tensor spaces

### PR DESCRIPTION
**Pull request by jorgensd**
To be able to easily extract blocks from `ufl.MixedFuntionSpaces`.

Motivation:
Instead of writing out every block of `a` and `L` by hand, as
```python
    a00 = ufl.inner(ufl.grad(u), ufl.grad(v)) * dx
    a01 = ufl.inner(c, v) * dx
    a10 = ufl.inner(u, d) * dx

    L0 = ufl.inner(f, v) * dx + ufl.inner(g, v) * ufl.ds
    L1 = ufl.inner(zero, d) * dx

    a = dolfinx.fem.form([[a00, a01], [a10, None]], dtype=dtype)
    L = dolfinx.fem.form([L0, L1], dtype=dtype)

```
one can now do
```python
    dx = ufl.Measure("dx", domain=mesh)
    f = -ufl.div(ufl.grad(u_ex))
    n = ufl.FacetNormal(mesh)
    g = ufl.dot(ufl.grad(u_ex), n)
    a = ufl.inner(ufl.grad(u), ufl.grad(v)) * dx
    a += ufl.inner(c, v) * dx
    a += ufl.inner(u, d) * dx
    L = ufl.inner(f, v) * dx + ufl.inner(g, v) * ufl.ds
    L += ufl.inner(zero, d) * dx

    a_blocked = ufl.extract_blocks(a)
    L_blocked = ufl.extract_blocks(L)

    a = dolfinx.fem.form(a_blocked, dtype=dtype)
    L = dolfinx.fem.form(L_blocked, dtype=dtype)
```

It also fixes `extract_blocks` for tensors, and properly document this function, as to what happens if `i` or `j` is not provided.
Now it extracts the blocks as a scalar, vector or tensor.

### CODE REVIEW

File: ufl/finiteelement.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def components(self) -> _typing.Dict[_typing.Tuple[int, ...], int]:
         offset = 0
         c_offset = 0
         for e in self.sub_elements:
-            for i, j in enumerate(np.ndindex(e.value_shape)):
+            for i, j in enumerate(np.ndindex(e.reference_value_shape)):
```
Comment (mscroggs) [inline ufl/finiteelement.py:151]: This change shouldn't be in this PR
```suggestion
            for i, j in enumerate(np.ndindex(e.value_shape)):
```

**Review by mscroggs — APPROVED** [review]

File: ufl/finiteelement.py (hunk 1 above)
Comment (jorgensd) [reply ufl/finiteelement.py:151]: I'll revert it once 307 is merged.

File: ufl/algorithms/formsplitter.py [hunk 2]
```diff
@@ -92,33 +84,60 @@ def multi_index(self, obj):
...
+    If arity is 0, returns the form.
+    If arity is 1, return the ith block. If ``i`` is ``None``, return all blocks.
+    If arity is 2, return the ``(i,j)`` entry. If ``j`` is ``None``, return the ith row.
+
+    If neither `i` nor `j` are set, return all blocks (as a scalar, vector or tensor).
+
+    Args:
+        form: A form
+        i: Index of the block to extract. If set to ``None``, ``j`` must be None.
+        j: Index of the block to extract. If set to ``None``, ``i`` must be None.
```
Comment (cdaversin) [inline ufl/algorithms/formsplitter.py:99]: Not consistent with : "If ``j`` is ``None``, return the ith row."

Comment (jorgensd) [inline ufl/algorithms/formsplitter.py:99]:
```suggestion
        j: Index of the block to extract.
```

File: ufl/algorithms/formsplitter.py [hunk 3]
```diff
@@ -131,10 +150,9 @@ def extract_blocks(form, i=None, j=None):
     except TypeError:
         # Only one form returned
         forms_tuple = (forms,)
-
     if i is not None:
         if arity > 1 and j is not None:
-            return forms_tuple[i * len(parts) + j]
+            return forms_tuple[i][j]
         else:
             return forms_tuple[i]
```
Comment (cdaversin) [inline ufl/algorithms/formsplitter.py:157]: Might be useful to have a warning if we end up there with j != None

File: ufl/algorithms/formsplitter.py [hunk 4]
```diff
@@ -92,33 +84,60 @@ def multi_index(self, obj):
...
+                forms.append(tuple(form_i))
+            return tuple(forms)
         else:
-            # If standard element, extract only part
-            parts.append(a.part())
-    parts = tuple(sorted(set(parts)))
-    for pi in parts:
+            return fs.split(form, i, j)
+
+    # If mixed function space, each argument has sub-elements
```
Comment (cdaversin) [inline ufl/algorithms/formsplitter.py:128]: Maybe some safety check on i and j depending on number of components in the MixedFunctionSpace ?

**Review by cdaversin — APPROVED** [review]
Looks good to me :)

**Comment by jpdean** [comment]
Nice! This'll be really handy for coupled problems. Do we have a test?

**Comment by jorgensd** [comment]
> Nice! This'll be really handy for coupled problems. Do we have a test?

I think we can adapt the stokes demo.

I use this in many of my own works lately.

**Pull request merged**